### PR TITLE
stage #2, Collections -> lesson #4 Queue initial commit

### PR DESCRIPTION
**Pull request by AlehDrazdou**
Queue

### CODE REVIEW

File: stage #2/module #3 Collections/lesson #4 Queue/queue.md [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+* Deque
+  * LinkedList
+  * ArrayDeque
+* Materials
+
+
+## Hierarchy. Queue.
+![](./media/queue_hierarchy.png)
+
+The Java **Queue** interface, _java.util.Queue_ represents a data structure designed to have elements inserted at the end of the queue, and elements removed from the beginning of the queue. it follows the FIFO or the First-In-First-Out principle.
```
Comment (BogdanovaDiana) [inline stage #2/module #3 Collections/lesson #4 Queue/queue.md:14]: I believe a capital I is needed here (It follows)

File: stage #2/module #3 Collections/lesson #4 Queue/queue.md [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+Method | Second Description
+------------ | -------------
+**boolean add(object)** | Inserts the specified element into this queue and returns true upon success
+**boolean offer(object)** | Insert the specified element into this queue
+**Object remove()** | Retrieves and removes the head of this queue
+**Object poll()** | Retrieves and removes the head of this queue, or returns null if this queue is empty
+**Object element()** | Retrieves, but does not remove, the head of this queue
+**Object peek()** | Retrieves, but does not remove, the head of this queue, or returns null if this queue is empty
+
+##PriorityQueue
```
Comment (BogdanovaDiana) [inline stage #2/module #3 Collections/lesson #4 Queue/queue.md:29]: Missing space, no formatting applied (## PriorityQueue)

File: stage #2/module #3 Collections/lesson #4 Queue/queue.md [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        Kristina
+        Valery
+        Ralph
+        after removing two elements:
+        Kristina
+        Ralph
+        Valery
+```
+
+###Features
```
Comment (BogdanovaDiana) [inline stage #2/module #3 Collections/lesson #4 Queue/queue.md:70]: Missing space, no formatting applied (## Features)

File: stage #2/module #3 Collections/lesson #4 Queue/queue.md [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+* Objects added to _PriorityQueue_ MUST be comparable
+* The objects of the priority queue are **ordered by default in natural order**
+* A _Comparator_ can be used for custom ordering of objects in the queue
+* The **head** of the priority queue is the **least** element based on the natural ordering or comparator based ordering. When we poll the queue, it returns the head object from the queue
+* If multiple objects are present of same priority - it can poll any one of them randomly
+* _PriorityQueue_ is not thread safe. Use _PriorityBlockingQueue_ in concurrent environment
+* It provides **O(log(n))** time for add and poll methods
+
+
+##Deque
```
Comment (BogdanovaDiana) [inline stage #2/module #3 Collections/lesson #4 Queue/queue.md:82]: Missing space, no formatting applied (## Deque)

File: stage #2/module #3 Collections/lesson #4 Queue/queue.md [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+The **Deque** interface present in _java.util_ package is a subtype of the **Queue** interface. The **Deque** is related to the _double-ended_ queue that supports addition or removal of elements from either end of the data structure. **Deque** is the acronym for _Double Ended Queue_.
+
+The Java **Deque** interface extends the Java Queue interface. That means that you can use all the Java **Queue** methods when working with a **Deque**.
+
+Since Java **Deque** is an interface you need to instantiate a concrete implementation of the interface in order to use it:
+* java.util.LinkedList
+* java.util.ArrayDeque
+
+###Features
```
Comment (BogdanovaDiana) [inline stage #2/module #3 Collections/lesson #4 Queue/queue.md:91]: Missing space, no formatting applied (### Features)

File: stage #2/module #3 Collections/lesson #4 Queue/queue.md [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+* **Deque** interface supports resizable arrays that can grow as required
+* Array deques _do not allow the use of NULL values_
+* **Deque** does not support concurrent access by more than one thread
+* **Deque** _is not thread-safe_ unless an external synchronization is provided
+
+##LinkedList
```
Comment (BogdanovaDiana) [inline stage #2/module #3 Collections/lesson #4 Queue/queue.md:97]: Missing space, no formatting applied (## LinkedList)

File: stage #2/module #3 Collections/lesson #4 Queue/queue.md [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+        System.out.println("Deque after removing "
+                           + "first and last: "
+                           + linkedListDeque);
+                           
+        Output:
+        [Element 6 (Head), Element 4 (Head), Element 2 (Head), Element 1 (Tail), Element 3 (Tail), Element 5 (Tail)]
+        Deque after removing first and last: [Element 4 (Head), Element 2 (Head), Element 1 (Tail), Element 3 (Tail)]                   
+```
+
+##ArrayDeque
```
Comment (BogdanovaDiana) [inline stage #2/module #3 Collections/lesson #4 Queue/queue.md:131]: Missing space, no formatting applied (## ArrayDeque)

File: stage #2/module #3 Collections/lesson #4 Queue/queue.md [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+    John
+    Aleh
+    Valery
+    Misha
+    After pollLast() Traversal...
+    John
+    Aleh
+    Valery
+```
+##Materials
```
Comment (BogdanovaDiana) [inline stage #2/module #3 Collections/lesson #4 Queue/queue.md:161]: Missing space, no formatting applied (## Materials)

File: stage #2/module #3 Collections/lesson #4 Queue/queue.md [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+
+Java , Head First book
+
+Thinking in Java, Eckel book
+
+https://docs.oracle.com/javase/8/docs/api/java/util/Queue.html
+
```
Comment (KristinaPotapovich) [inline stage #2/module #3 Collections/lesson #4 Queue/queue.md:168]: looks like the same links

**Pull request merged**